### PR TITLE
[AArch64][GISel] Always fold G_SHL into addressing mode where possible, unless the subtarget has addr-lsl-slow-14

### PR DESCRIPTION
**Pull request by Him188**
Before this patch, we fold G_SHL into addressing mode lsl only when there is exactly one usage, or all the usages are memory ops, or we are optimizing for size. However, lsl is free on all aarch64 targets except those with FeatureAddrLSLSlow14. 

This patch uses this fact and always folds G_SHL into lsl for memory ops, with exceptions for FeatureAddrLSLSlow14.

This patch also fixes GISel 15% regression in TSVC kernel s482, and brings regression in s291 from 20% to 10%.

### CODE REVIEW

**Comment by davemgreen** [comment]
See #86894 for how this should work. Some addressing modes are effectively free, so I don't believe it should matter of there are other uses. There are some target features that control what is considered free (as-in `FIXME: Consider checking HasAddrLSLSlow14 and HasALULSLFast as appropriate.`)

**Comment by Him188** [comment]
@davemgreen Does the updated solution solves your concerns? I will polish it if we are on the right way.

**Comment by davemgreen** [comment]
I will have to check the details, but yeah I think this looks decent. It would be good to have a .ll test file that tests various cases with GISel and SDAG run lines to check that they are folding at the same times.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 1]
```diff
@@ -6717,19 +6722,72 @@ AArch64InstructionSelector::selectNegArithImmed(MachineOperand &Root) const {
...
+    // Fold G_PTR_ADD if its offset operand can be folded
+    if (MI.getOpcode() == AArch64::G_PTR_ADD) {
+      MachineInstr *OffsetInst =
+          getDefIgnoringCopies(MI.getOperand(2).getReg(), MRI);
+
+      // Note, we already know G_PTR_ADD is used by at least two instructions.
+      // If we are also sure about whether folding is beneficial or not,
+      // return the result.
+      if (const auto Worth = isWorthFoldingIntoAddrMode(*OffsetInst, MRI);
+          Worth.has_value())
```
Comment (arsenm) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:6785]: The .has_value check is redundant

Comment (tschuett) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:6785]: There are more oddities, but let's first wait for convergence on the solution.

Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp]: done

**Comment by Him188** [comment]
@davemgreen I have included one .ll test, part of which is copied from SDAG `aarch64-fold-lslfast.ll`. Also added MIR tests for s128 and stores.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 2]
```diff
@@ -6717,19 +6722,70 @@ AArch64InstructionSelector::selectNegArithImmed(MachineOperand &Root) const {
   return select12BitValueWithLeftShift(Immed);
 }
 
+/// Checks if we are sure that folding MI into load/store addressing mode is
+/// beneficial or not.
+///
+/// Returns:
+/// - true if fodling MI would be beneficial.
```
Comment (tschuett) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:6729]: folding

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 3]
```diff
@@ -6717,19 +6722,70 @@ AArch64InstructionSelector::selectNegArithImmed(MachineOperand &Root) const {
...
   if (MRI.hasOneNonDBGUse(DefReg) ||
       MI.getParent()->getParent()->getFunction().hasOptSize())
     return true;
 
-  // FIXME: Consider checking HasAddrLSLSlow14 and HasALULSLFast as
-  // appropriate.
+  if (IsAddrOperand) {
+    // If we are already sure that folding MI is good or bad, return the result.
+    if (const auto Worth = isWorthFoldingIntoAddrMode(MI, MRI))
+      return Worth.value();
```
Comment (tschuett) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:6773]: optionals are odd, see https://en.cppreference.com/w/cpp/utility/optional.

```
return *Worth;
```

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 4]
```diff
@@ -6717,19 +6722,70 @@ AArch64InstructionSelector::selectNegArithImmed(MachineOperand &Root) const {
...
+    // Fold G_PTR_ADD if its offset operand can be folded
+    if (MI.getOpcode() == AArch64::G_PTR_ADD) {
+      MachineInstr *OffsetInst =
+          getDefIgnoringCopies(MI.getOperand(2).getReg(), MRI);
+
+      // Note, we already know G_PTR_ADD is used by at least two instructions.
+      // If we are also sure about whether folding is beneficial or not,
+      // return the result.
+      if (const auto Worth = isWorthFoldingIntoAddrMode(*OffsetInst, MRI))
+        return Worth.value();
```
Comment (tschuett) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:6784]:
```
return *Worth;
```

**Comment by aemerson** [comment]
You can pre-commit the tests (don't have to have a PR) so that this change shows the code differences more clearly.

**Comment by Him188** [comment]
> You can pre-commit the tests

Nice suggestion. I have prepared one test commit before the code commit. You can review the second commit https://github.com/llvm/llvm-project/pull/96603/commits/c6ec76e712ad7d618e62dc72e4f64a0d03e2644d to see what's changed.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp (hunk 4 above)
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp]: done

File: llvm/test/CodeGen/AArch64/GlobalISel/select-fold-lslfast.ll [hunk 5]
```diff
@@ -12,17 +12,16 @@ declare void @foo()
 define i16 @halfword(ptr %ctx, i32 %xor72) nounwind {
```
Comment (davemgreen) [inline llvm/test/CodeGen/AArch64/GlobalISel/select-fold-lslfast.ll:12]: This test can be moved up a folder, if it's not just gisel. I would add the SDAG lines first too.

Can you remove the calls? It should still be possible to have different uses if it stores a different value.

Comment (Him188) [reply llvm/test/CodeGen/AArch64/GlobalISel/select-fold-lslfast.ll:12]: This was actually coped from `AArch64/aarch64-fold-lslfast.ll`. Let's add GISEL RUN lines to this file and delete the new file.

Comment (Him188) [reply llvm/test/CodeGen/AArch64/GlobalISel/select-fold-lslfast.ll]: The calls are originally there so I would prefer keeping them. Instead I added a new case `multi_use_half_word` to show multiple uses with stores (and without call).

**Review by davemgreen — COMMENTED** [review]
Thanks. I think this looks OK to me.

**Review by davemgreen — APPROVED** [review]
Thanks. LGTM

**Review by aemerson — APPROVED** [review]
Would be good to understand why we still have differences to SDAG in the tests at some point.

**Comment by Him188** [comment]
Thanks. NFC test cases merged separately. Will merge this after successful CI.

**Pull request merged**